## Read the data 

In [1]:
import pandas as pd 
import numpy as np 
pd.options.display.max_columns = 999 

In [2]:
loans_2007 = pd.read_csv('./Data/loans_2007.csv')
loans_2007.drop_duplicates()

/Users/bholaprasad/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3057: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
5                                              1075269  1311441.0     5000.0   
6                                              1069639  1304742.0     7000.0   
7                                              1072053  1288686.0     3000.0   
8                                              1071795  1306957.0     5600.0   
9                                              1071570  1306721.0     5375.0   
10                                             1070078  1305201.0     6500.0   
11                                             1069908  1305008.0    12000.0   
12                                             1064687  1298717.0     9000.0   
13                                             1069866  1304956.0     3000.0   
14                                             1069057  1303503.0    10000.0   
15                                             1069759  1304871.0     1000.0   
16                                             1065775  1299699.0    10000.0   
17                                             1069971  1304884.0     3600.0   
18                                             1062474  1294539.0     6000.0   
19                                             1069742  1304855.0     9200.0   
20                                             1069740  1284848.0    20250.0   
21                                             1039153  1269083.0    21000.0   
22                                             1069710  1304821.0    10000.0   
23                                             1069700  1304810.0    10000.0   
24                                             1069559  1304634.0     6000.0   
25                                             1069697  1273773.0    15000.0   
26                                             1069800  1304679.0    15000.0   
27                                             1069657  1304764.0     5000.0   
28                                             1069799  1304678.0     4000.0   
29                                             1047704  1278806.0     8500.0   
...                                                ...        ...        ...   
42508                                            91175    91170.0     6000.0   
42509                                            91126    91067.0     5350.0   
42510                                            91023    70879.0     1900.0   
42511                                            90106    90090.0    10000.0   
42512                                            89258    80039.0     2000.0   
42513                                            88637    88629.0     6000.0   
42514                                            88046    88023.0     4400.0   
42515                                            85961    85923.0     1200.0   
42516                                            85818    85802.0     5000.0   
42517                                            85781    85727.0     1400.0   
42518                                            85675    85667.0     1000.0   
42519                                            84670    79576.0     5000.0   
42520                                            84098    84091.0     2500.0   
42521                                            83979    83974.0     3000.0   
42522                                            83489    83471.0     2600.0   
42523                                            83185    83132.0     1000.0   
42524                                            74014    73890.0     6450.0   
42525                                            76629    76623.0     1275.0   
42526                                            810

In [3]:
loans_2007.shape

(42538, 52)

## First group of features

The Dataframe contains many columns and can be cumbersome to try to explore all at once. Let's break up the columns into 3 groups of 18 columns and use the [data dictionary](https://docs.google.com/spreadsheets/d/191B2yJ4H1ZPXq0_ByhUgWMFZOYem5jFz0Y3by_7YBY4/edit#gid=2081333097) to become familiar with what each column represents. As you understand each feature, you want to pay attention to any features that:

* leak information from the future (after the loan has already been funded)
* don't affect a borrower's ability to pay back a loan (e.g. a randomly generated ID value by Lending Club)
* formatted poorly and need to be cleaned up
* require more data or a lot of processing to turn into a useful feature
* contain redundant information



The features that needs to be removed:

* id: randomly generated field by Lending Club for unique identification purposes only
* member_id: also a randomly generated field by Lending Club for unique identification purposes only
* funded_amnt: leaks data from the future (after the loan is already started to be funded)
* funded_amnt_inv: also leaks data from the future (after the loan is already started to be funded)
* grade: contains redundant information as the interest rate column (int_rate)
* sub_grade: also contains redundant information as the interest rate column (int_rate)
* emp_title: requires other data and a lot of processing to potentially be useful
* issue_d: leaks data from the future (after the loan is already completed funded)
Recall that Lending Club assigns a grade and a sub-grade based on the borrower's interest rate. While the grade and sub_grade values are categorical, the int_rate column contains continuous values, which are better suited for machine learning.

Let's now drop these columns from the Dataframe before moving onto the next group of columns.

In [4]:
loans_2007 = loans_2007.drop(['id','member_id','funded_amnt','funded_amnt_inv','grade','sub_grade',
                              'emp_title','issue_d'], axis=1)

## Second group of features

we need to drop the following columns:

* zip_code: redundant with the addr_state column since only the first 3 digits of the 5 digit zip code are visible (which only can be used to identify the state the borrower lives in)
* out_prncp: leaks data from the future, (after the loan already started to be paid off)
* out_prncp_inv: also leaks data from the future, (after the loan already started to be paid off)
* total_pymnt: also leaks data from the future, (after the loan already started to be paid off)
* total_pymnt_inv: also leaks data from the future, (after the loan already started to be paid off)
* total_rec_prncp: also leaks data from the future, (after the loan already started to be paid off)

In [5]:
loans_2007 = loans_2007.drop(['zip_code', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv',
                              'total_rec_prncp'], axis=1)

## Third group of features

In the last group of columns, we need to drop the following columns:

* total_rec_int: leaks data from the future, (after the loan already started to be paid off),
* total_rec_late_fee: also leaks data from the future, (after the loan already started to be paid off),
* recoveries: also leaks data from the future, (after the loan already started to be paid off),
* collection_recovery_fee: also leaks data from the future, (after the loan already started to be paid off),
* last_pymnt_d: also leaks data from the future, (after the loan already started to be paid off),
* last_pymnt_amnt: also leaks data from the future, (after the loan already started to be paid off).
    
All of these columns leak data from the future, meaning that they're describing aspects of the loan after it's already been fully funded and started to be paid off by the borrower.

In [6]:
loans_2007 = loans_2007.drop(['total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 
                              'last_pymnt_d', 'last_pymnt_amnt'], axis=1)

In [7]:
loans_2007.head()

loan_amnt        term int_rate  installment emp_length home_ownership  \
0     5000.0   36 months   10.65%       162.87  10+ years           RENT   
1     2500.0   60 months   15.27%        59.83   < 1 year           RENT   
2     2400.0   36 months   15.96%        84.33  10+ years           RENT   
3    10000.0   36 months   13.49%       339.31  10+ years           RENT   
4     3000.0   60 months   12.69%        67.79     1 year           RENT   

   annual_inc verification_status  loan_status pymnt_plan         purpose  \
0     24000.0            Verified   Fully Paid          n     credit_card   
1     30000.0     Source Verified  Charged Off          n             car   
2     12252.0        Not Verified   Fully Paid          n  small_business   
3     49200.0     Source Verified   Fully Paid          n           other   
4     80000.0     Source Verified      Current          n           other   

                  title addr_state    dti  delinq_2yrs earliest_cr_line  \
0              Computer         AZ  27.65          0.0         Jan-1985   
1                  bike         GA   1.00          0.0         Apr-1999   
2  real estate business         IL   8.72          0.0         Nov-2001   
3              personel         CA  20.00          0.0         Feb-1996   
4              Personal         OR  17.94          0.0         Jan-1996   

   inq_last_6mths  open_acc  pub_rec  revol_bal revol_util  total_acc  \
0             1.0       3.0      0.0    13648.0      83.7%        9.0   
1             5.0       3.0      0.0     1687.0       9.4%        4.0   
2             2.0       2.0      0.0     2956.0      98.5%       10.0   
3             1.0      10.0      0.0     5598.0        21%       37.0   
4             0.0      15.0      0.0    27783.0      53.9%       38.0   

  initial_list_status last_credit_pull_d  collections_12_mths_ex_med  \
0                   f           Jun-2016                         0.0   
1                   f           Sep-2013                         0.0   
2                   f           Jun-2016                         0.0   
3                   f           Apr-2016                         0.0   
4                   f           Jun-2016                         0.0   

   policy_code application_type  acc_now_delinq  chargeoff_within_12_mths  \
0          1.0       INDIVIDUAL             0.0                       0.0   
1          1.0       INDIVIDUAL             0.0                       0.0   
2          1.0       INDIVIDUAL             0.0                       0.0   
3          1.0       INDIVIDUAL             0.0                       0.0   
4          1.0       INDIVIDUAL             0.0                       0.0   

   delinq_amnt  pub_rec_bankruptcies  tax_liens  
0          0.0                   0.0        0.0  
1          0.0                   0.0        0.0  
2          0.0                   0.0        0.0  
3          0.0                   0.0        0.0  
4          0.0                   0.0        0.0

In [8]:
loans_2007.shape

(42538, 32)

## Target Column

Just by becoming familiar with the columns in the dataset, we were able to reduce the number of columns from 52 to 32 columns. We now need to decide on a target column that we want to use for modeling.

We should use the loan_status column, since it's the only column that directly describes if a loan was paid off on time, had delayed payments, or was defaulted on the borrower. Currently, this column contains text values and we need to convert it to a numerical one for training a model. Let's explore the different values in this column and come up with a strategy for converting the values in this column.

In [9]:
loans_2007['loan_status'].value_counts()

Fully Paid                                             33136
Charged Off                                             5634
Does not meet the credit policy. Status:Fully Paid      1988
Current                                                  961
Does not meet the credit policy. Status:Charged Off      761
Late (31-120 days)                                        24
In Grace Period                                           20
Late (16-30 days)                                          8
Default                                                    3
Name: loan_status, dtype: int64

From the investor's perspective, we're interested in trying to predict which loans will be paid off on time and which ones won't be. Only the Fully Paid and Charged Off values describe the final outcome of the loan. The other values describe loans that are still on going and where the jury is still out on if the borrower will pay back the loan on time or not. While the Default status resembles the Charged Off status, in Lending Club's eyes, loans that are charged off have essentially no chance of being repaid while default ones have a small chance. You can read about the difference here.

Since we're interested in being able to predict which of these 2 values a loan will fall under, we can treat the problem as a binary classification one. Let's remove all the loans that don't contain either Fully Paid and Charged Off as the loan's status and then transform the Fully Paid values to 1 for the positive case and the Charged Off values to 0 for the negative case. 

Lastly, one thing we need to keep in mind is the class imbalance between the positive and negative cases. While there are 33,136 loans that have been fully paid off, there are only 5,634 that were charged off. This class imbalance is a common problem in binary classification and during training, the model ends up having a strong bias towards predicting the class with more observations in the training set and will rarely predict the class with less observations. The stronger the imbalance, the more biased the model becomes. There are a few different ways to tackle this class imbalance, which we'll explore later.

## Binary Classification

In [10]:
loans_2007 = loans_2007[(loans_2007['loan_status'] == 'Fully Paid') | (loans_2007['loan_status'] == 'Charged Off')]

status_replace = {
    'loan_status' : {
        'Fully Paid': 1,
        'Charged Off': 0,
    }
}

loans_2007 = loans_2007.replace(status_replace)

## Removing single value columns

In [11]:
drop_columns = []

for col in loans_2007.columns:
    non_null = loans_2007[col].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    
    if num_true_unique == 1:
        drop_columns.append(col)
        
loans_2007 = loans_2007.drop(drop_columns, axis=1)
print(drop_columns)

['pymnt_plan', 'initial_list_status', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'chargeoff_within_12_mths', 'delinq_amnt', 'tax_liens']


## Save the cleaned data set

In [12]:
loans_2007.to_csv('./Data/filtered_loans_2007.csv', index=False)